### ISSUE TITLE
Data race in delete predicate test

### ISSUE THREAD

**Issue by pawanrawal** [dgraph-io/dgraph#1544]
This is the test in `main_test.go`.

```
==================
WARNING: DATA RACE
Read at 0x00c4200b01e4 by goroutine 1090:
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*valueLog).Read()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/value.go:824 +0x194
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*KV).yieldItemValue()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/kv.go:419 +0x5f2
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*KVItem).prefetchValue()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/iterator.go:87 +0xc2
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*Iterator).fill.func1()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/iterator.go:253 +0x38

Previous write at 0x00c4200b01e4 by goroutine 2097:
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*valueLog).write.func1()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/value.go:750 +0x4b2
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*valueLog).write()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/value.go:801 +0x88d
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*KV).writeRequests()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/kv.go:603 +0x2b3
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*KV).doWrites.func1()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/kv.go:643 +0x6f

Goroutine 1090 (running) created at:
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*Iterator).fill()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/iterator.go:251 +0x3e6
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*Iterator).prefetch()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/iterator.go:278 +0x273
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*Iterator).Seek()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/iterator.go:302 +0x641
  github.com/dgraph-io/dgraph/posting.RebuildIndex()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/posting/index.go:706 +0x449
  github.com/dgraph-io/dgraph/worker.(*node).rebuildOrDelIndex()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/worker/index.go:45 +0x243
  github.com/dgraph-io/dgraph/worker.runSchemaMutation()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/worker/mutation.go:144 +0x901
  github.com/dgraph-io/dgraph/worker.(*node).processSchemaMutations()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/worker/draft.go:341 +0x154
  github.com/dgraph-io/dgraph/worker.(*scheduler).schedule()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/worker/scheduler.go:114 +0x20a
  github.com/dgraph-io/dgraph/worker.(*node).processApplyCh()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/worker/draft.go:406 +0x45f

Goroutine 2097 (finished) created at:
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*KV).doWrites()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/kv.go:692 +0x44e
==================
==================
WARNING: DATA RACE
Read at 0x00c4200b01e4 by goroutine 2123:
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*valueLog).Read()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/value.go:824 +0x194
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*KV).yieldItemValue()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/kv.go:419 +0x5f2
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*KVItem).prefetchValue()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/iterator.go:87 +0xc2
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*Iterator).fill.func1()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/iterator.go:253 +0x38

Previous write at 0x00c4200b01e4 by goroutine 2097:
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*valueLog).write.func1()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/value.go:750 +0x4b2
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*valueLog).write()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/value.go:801 +0x88d
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*KV).writeRequests()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/kv.go:603 +0x2b3
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*KV).doWrites.func1()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/kv.go:643 +0x6f

Goroutine 2123 (running) created at:
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*Iterator).fill()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/iterator.go:251 +0x3e6
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*Iterator).prefetch()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/iterator.go:278 +0x273
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*Iterator).Seek()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/iterator.go:302 +0x641
  github.com/dgraph-io/dgraph/posting.RebuildIndex()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/posting/index.go:706 +0x449
  github.com/dgraph-io/dgraph/worker.(*node).rebuildOrDelIndex()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/worker/index.go:45 +0x243
  github.com/dgraph-io/dgraph/worker.runSchemaMutation()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/worker/mutation.go:144 +0x901
  github.com/dgraph-io/dgraph/worker.(*node).processSchemaMutations()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/worker/draft.go:341 +0x154
  github.com/dgraph-io/dgraph/worker.(*scheduler).schedule()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/worker/scheduler.go:114 +0x20a
  github.com/dgraph-io/dgraph/worker.(*node).processApplyCh()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/worker/draft.go:406 +0x45f

Goroutine 2097 (finished) created at:
  github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*KV).doWrites()
      /home/pawan/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/kv.go:692 +0x44e
==================
2017/09/27 11:28:45 mutation.go:160: Done schema update predicate:"friend" value_type:9 directive:INDEX tokenizer:"term" explicit:true 
2017/09/27 11:28:45 schema.go:78: Setting schema for attr name: uid, tokenizer: [], directive: REVERSE, count: false
2017/09/27 11:28:45 mutation.go:160: Done schema update predicate:"name" value_type:7 directive:REVERSE explicit:true 
--- FAIL: TestDeletePredicate (6.10s)
	testing.go:699: race detected during execution of test
```

**Comment by pawanrawal** [comment]
This gets fixed by https://github.com/dgraph-io/badger/pull/248

**Comment by manishrjain** [comment]
Nice fix!

**Comment by pawanrawal** [comment]
This is fixed after upgrading Badger in 7b64e51b2ca2a3c9e9a58db638b1d96ae9479f9b
